File: bindings/csharp/examples/BuilderPatternExample.cs
using System;
using Psyne;

namespace Psyne.Examples
{
    /// <summary>
    /// Demonstrates the fluent builder API for channel creation and configuration.
    /// </summary>
    public class BuilderPatternExample
    {
        public static void Run()
        {
            Console.WriteLine("=== Builder Pattern Example ===");

            Psyne.Initialize();

            try
            {
                // Example 1: Basic memory channel with custom buffer size
                Console.WriteLine("1. Creating memory channel with builder...");
                using var memoryChannel = Psyne.CreateChannel()
                    .Memory("builder-example")
                    .WithBufferSize(5, SizeUnit.MB)
                    .SingleProducerSingleConsumer()
                    .SingleType()
                    .WithMetrics()
                    .Build();

                Console.WriteLine($"Memory channel created: {memoryChannel.Uri}");
                Console.WriteLine($"Buffer size: {memoryChannel.BufferSize:N0} bytes");

                // Test the memory channel
                memoryChannel.Send("Memory channel test message");
                using var msg1 = memoryChannel.Receive();
                Console.WriteLine($"Received: {msg1?.GetString()}");

                // Example 2: TCP channel with compression
                Console.WriteLine("\n2. Creating TCP channel with compression...");
                using var tcpChannel = Psyne.CreateChannel()
                    .Tcp("localhost", 8080)
                    .WithBufferSize(2, SizeUnit.MB)
                    .MultipleProducerMultipleConsumer()
                    .MultiType()
                    .WithLz4Compression()
                    .WithMetrics()
                    .Build();

                Console.WriteLine($"TCP channel created: {tcpChannel.Uri}");

                // Example 3: Unix socket channel with custom compression
                Console.WriteLine("\n3. Creating Unix socket channel with custom compression...");
                var customCompression = new CompressionConfig
                {
                    Type = CompressionType.Zstd,
                    Level = 6,
                    MinSizeThreshold = 512,
                    EnableChecksum = true
                };

                using var unixChannel = Psyne.CreateChannel()
                    .UnixSocket("/tmp/psyne-builder-example.sock")
                    .WithBufferSize(1024, SizeUnit.KB)
                    .MultipleProducerSingleConsumer()
                    .WithCompression(customCompression)
                    .Build();

                Console.WriteLine($"Unix channel created: {unixChannel.Uri}");

                // Example 4: UDP multicast channel with Snappy compression
                Console.WriteLine("\n4. Creating UDP multicast channel...");
                using var udpChannel = Psyne.CreateChannel()
                    .UdpMulticast("239.1.1.1", 8080)
                    .WithBufferSize(512, SizeUnit.KB)
                    .SingleProducerMultipleConsumer()
                    .WithSnappyCompression()
                    .Build();

                Console.WriteLine($"UDP channel created: {udpChannel.Uri}");

                // Example 5: Demonstrate different compression methods
                Console.WriteLine("\n5. Testing different compression methods...");

                var compressionConfigs = new[]
                {
                    ("LZ4", CompressionConfig.Lz4()),
                    ("Zstd", CompressionConfig.Zstd()),
                    ("Snappy", CompressionConfig.Snappy())
                };

                foreach (var (name, config) in compressionConfigs)
                {
                    using var compressedChannel = Psyne.CreateChannel()
                        .Memory($"compression-test-{name.ToLower()}")
                        .WithBufferSize(1, SizeUnit.MB)
                        .WithCompression(config)
                        .WithMetrics()
                        .Build();

                    Console.WriteLine($"Created {name} compressed channel: {compressedChannel.Uri}");

                    // Send a large message to test compression
                    var largeMessage = new string('A', 10000); // 10KB of 'A's
                    compressedChannel.Send(largeMessage);

                    using var received = compressedChannel.Receive();
                    if (received != null && received.GetString() == largeMessage)
                    {
                        Console.WriteLine($"  {name} compression test passed");
                        
                        var metrics = compressedChannel.GetMetrics();
                        Console.WriteLine($"  Messages sent: {metrics.MessagesSent}");
                        Console.WriteLine($"  Bytes sent: {metrics.BytesSent:N0}");
                    }
                }

                // Example 6: Method chaining variations
                Console.WriteLine("\n6. Method chaining variations...");

                // Variation 1: Explicit methods
                using var explicit1 = Psyne.CreateChannel()
                    .Memory("explicit-1")
                    .SingleProducerSingleConsumer()
                    .SingleType()
                    .Build();

                // Variation 2: Shorthand methods
                using var explicit2 = Psyne.CreateChannel()
                    .Memory("explicit-2")
                    .WithMode(ChannelMode.SingleProducerSingleConsumer)
                    .WithType(ChannelType.Single)
                    .Build();

                Console.WriteLine("Both variations created successfully");

                Console.WriteLine("\nBuilder pattern example completed successfully!");
            }
            catch (PsyneException ex)
            {
                Console.WriteLine($"Psyne error: {ex.ErrorCode} - {ex.Message}");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Unexpected error: {ex.Message}");
            }
            finally
            {
                Psyne.Cleanup();
            }
        }
    }
}

File: bindings/swift/Examples/BuilderPatternExample.swift
import Foundation
import Psyne

/// Example demonstrating the builder pattern for channel creation
@main
struct BuilderPatternExample {
    static func main() async {
        do {
            // Initialize the Psyne library
            try Psyne.initialize()
            defer { Psyne.cleanup() }
            
            print("Psyne Swift Builder Pattern Example")
            print("Version: \(Psyne.version)")
            print()
            
            // Example 1: Memory channel with custom configuration
            print("Creating memory channel with custom configuration...")
            let memoryChannel = try Psyne.createChannel()
                .memory("custom_memory")
                .withBufferSize(2 * 1024 * 1024) // 2MB buffer
                .multipleProducersMultipleConsumers()
                .multipleMessageTypes()
                .withLZ4Compression()
                .build()
            
            print("Memory channel created: \(memoryChannel.uri)")
            print("Buffer size: \(try memoryChannel.getBufferSize()) bytes")
            
            // Example 2: TCP server channel
            print("\nCreating TCP server channel...")
            let tcpServer = try Psyne.createChannel()
                .tcpServer(port: 8080)
                .withBufferSize(512 * 1024) // 512KB buffer
                .singleProducerSingleConsumer()
                .singleMessageType()
                .withSnappyCompression()
                .build()
            
            print("TCP server created: \(tcpServer.uri)")
            
            // Example 3: Unix socket channel with compression
            print("\nCreating Unix socket channel...")
            let unixChannel = try Psyne.createChannel()
                .unixSocket("/tmp/psyne_example.sock")
                .withBufferSize(1024 * 1024) // 1MB buffer
                .multipleProducersSingleConsumer()
                .withZstdCompression()
                .build()
            
            print("Unix socket channel created: \(unixChannel.uri)")
            
            // Example 4: Custom URI with detailed configuration
            print("\nCreating channel with custom URI...")
            let customChannel = try Psyne.createChannel()
                .withUri("ipc://shared_memory_segment")
                .withBufferSize(4 * 1024 * 1024) // 4MB buffer
                .withMode(.mpmc)
                .withType(.multiType)
                .withCompression(CompressionConfig(
                    type: .zstd,
                    level: 9,
                    minSizeThreshold: 256,
                    enableChecksum: true
                ))
                .build()
            
            print("Custom channel created: \(customChannel.uri)")
            
            // Demonstrate sending messages through different channels
            print("\nTesting message sending...")
            
            // Send through memory channel
            try memoryChannel.sendText("Hello from memory channel!")
            
            // Send through custom channel
            let jsonData = [
                "message": "Hello from custom channel!",
                "timestamp": Date().timeIntervalSince1970,
                "channel": "custom"
            ]
            try customChannel.sendJSON(jsonData)
            
            // Send ML data through unix channel
            let mlData: [Float] = (0..<100).map { Float($0) * 0.1 }
            try unixChannel.sendFloatArray(mlData)
            
            // Receive messages
            print("\nReceiving messages...")
            
            if let message1 = try memoryChannel.receiveMessage(timeout: 1.0) {
                if let textMsg = message1.asTextMessage() {
                    print("Memory channel message: \(textMsg.text)")
                }
            }
            
            if let message2 = try customChannel.receiveMessage(timeout: 1.0) {
                if let jsonMsg = message2.asJSONMessage() {
                    print("Custom channel message: \(jsonMsg.json)")
                }
            }
            
            if let message3 = try unixChannel.receiveMessage(timeout: 1.0) {
                if let floatMsg = message3.asFloatArrayMessage() {
                    print("Unix channel message: \(floatMsg.values.count) float values")
                    print("First few values: \(Array(floatMsg.values.prefix(5)))")
                }
            }
            
            // Show metrics for each channel
            print("\nChannel Metrics:")
            
            try memoryChannel.setMetricsEnabled(true)
            let memoryMetrics = try memoryChannel.getMetrics()
            print("Memory Channel:")
            print("  Messages sent: \(memoryMetrics.messagesSent)")
            print("  Bytes sent: \(memoryMetrics.bytesSent)")
            
            try customChannel.setMetricsEnabled(true)
            let customMetrics = try customChannel.getMetrics()
            print("Custom Channel:")
            print("  Messages sent: \(customMetrics.messagesSent)")
            print("  Bytes sent: \(customMetrics.bytesSent)")
            
            try unixChannel.setMetricsEnabled(true)
            let unixMetrics = try unixChannel.getMetrics()
            print("Unix Channel:")
            print("  Messages sent: \(unixMetrics.messagesSent)")
            print("  Bytes sent: \(unixMetrics.bytesSent)")
            
            print("\nBuilder pattern example completed!")
            
        } catch {
            print("Error: \(error)")
        }
    }
}